### ISSUE TITLE
How to run unit tests under .../Node/core/tests/actions.js

### ISSUE THREAD

**Issue by rjgmail88** [Microsoft/BotBuilder#4214]
## Bot Info
* SDK Platform:  Node.js

## Issue Description
How to run the test cases under package ..../Node/core/tests/actions.js
I copied app the test cases from actions.js into a test file and modified the dialog names as per my code. But I get an error saying 
ReferenceError: describe is not defined

## Code Example
describe('actions', function() {
    this.timeout(5000);
    it('should launch dialog using a triggerAction()', function (done) { 
        var connector = new builder.ConsoleConnector();       
        var bot = new builder.UniversalBot(connector);
        bot.dialog('UserInfo', function (session, args) { 
            assert(session !== null);
            assert(args !== null);
            done();
        }).triggerAction({ matches: /UserInfo/i });
        connector.processMessage('what is my phone number');
    });

**Comment by microsoftly** [comment]
@rjgmail88 I know this isn't exactly the answer you're looking for, but I found pain doing exactly what you are now.

I made a package to help me address the headache of testing the bots, check it out! Several companies are using it in their CICD pipeline. I personally use it over the emulator most of the time because it is faster and more reliable. 

-> https://github.com/microsoftly/BotTester

**Comment by rjgmail88** [comment]
Thanks. I was able to run the code I mentioned by installing **mocha** module. But, the issue again is I have to run each test file by mocha {testFileNam}
Also the files under .../Node/core/tests/.. just shows general testing with each dialog defined in its own 
it(..... 
});
but it doesn't shows how to test out own custom dialogs and conversation flow.

**Comment by rjgmail88** [comment]
@microsoftly,
I went over your BotTester repo and here is the code for which I have a question.

{code}
 it('can test prompts', () => {
        bot.dialog('/', [(session) => {
            new Prompts.text(session, 'Hi there! Tell me something you like');
        }, (session, results) => {
            session.send(`${results.response} is pretty cool.`);
            new Prompts.text(session, 'Why do you like it?');
        }, (session) => session.send('Interesting. Well, that\'s all I have for now')]);

        return new BotTester(bot)
            .sendMessageToBot('Hola!', 'Hi there! Tell me something you like')
            .sendMessageToBot('The sky', 'The sky is pretty cool.', 'Why do you like it?')
            .sendMessageToBot('It\'s blue', 'Interesting. Well, that\'s all I have for now')
            .runTest();
    });
{code}
so here you are testing a dialog with prompt and BotTester you are making sure it works that way as prompt responses are designed. 

I have a bot service which is Cortana skill for booking outlook meetings ,giving people info from ms graph etc. Ex: here is my dialog conversation in text 

User: what is Rohit's phone number
bot:Rohit's phone number is Xxx-xxx-xxxx

How would I test it using your pattern ? As in do I need to redefine same dialog and test ? or is there a way to call my dialog from test file and test the responses. I have LUIS recognizer as well.

**Comment by microsoftly** [comment]
Several layers here.

To start, the framework does not make any assumptions about the bot. It can live in production or live in a completely mocked test world. If you have a way of mocking/setting user data/conversation flow at any point, you can use it and pick up using the bot-tester framework. If you do not, you can setup the scenario for the particular test case/dialog as if you were communicating with the bot (e.g. write the entire dialog flow and run it through the bot-tester).

You don't actually need to run assertions against the bot-tester, it can also move the state of the bot along as you need, allowing you to continue to setup the scenario you are running.

Within the scope of the test file, you do not need to modify any aspect of your bot code and it should simulate the bot running in an arbitrary environment just fine. 

If you have LUIS running, you probably don't want to pay to mock out the tests. Assuming that's the case, I would recommend using [nock](https://github.com/node-nock/nock) to mock requests the the LUIS service. You should be careful here to make sure you have the right response structured. Nock has a super awesome feature that let's you record actual requests and responses and spits out the code to mock it. That's probably the fastest way to get around LUIS.

TLDR;

``` javascript
var YourBot = require('your-bot');

/ /inside your test scope
    // setup any mock resources to start and seed db if necessary
    initializeLuisMocks();
    seedDbWithPhoneNumbers()
    const botInstance = new YourBot(new TestConnector());

    return new BotTester(botInstance)
        .sendMessageToBot('your initial convo setup goes here', 'you do not need to provide a response, though it would not hurt')
        // the actual test
        .sendMessageToBot('what is Rohit\'s phone number', 'Rohit\'s phone number is Xxx-xxx-xxxx')
        .runTest();
```

Added note: I built this for my personal use to test an extension I was making. You can see the extensive and advanced use cases [here](https://github.com/microsoftly/bot-handoff/blob/simplifyModel/test/providerTest.ts) (note that as of this writing, it is not on the master branch)

**Comment by rjgmail88** [comment]
@microsoftly, thanks for your detailed response. I am very new to bot-testing and just started writing some/exploring unit testing today. I am from java background so I was picturing it as I will pass an arg to by dialogs and just test if I'm getting expected responses back. But, looks like things are little different here in bot framework. Do you have any TEST for a dialog having trigger action by LUIS ?

I'am still having hard time figuring out the each piece here. For my skill 'User Info' there are bunch of things happing before I get response Rohit\'s phone number is Xxx-xxx-xxxx' in response.

{code}
var builder = require('botbuilder');
var personEntityConverter = require('../helpers/PersonEntityConverter.js')
var personInfoHelper = require('../helpers/PersonInfoHelper.js')

module.exports = function (bot, chatData) {

    bot.dialog('UserInfo', [
        function (session, args, next) {
            clearFlags(session)
             
           //LUIS resolves 'what is rohit's phone number' into 2 entities as follows 
            let personResults = personEntityConverter.convertUserInfoToObject(args.intent);
            //entity value is 'rohit'
             session.conversationData.PersonName = personResults.PersonName 
            //entity value is 'phone number'
             session.conversationData.PersonAttribute = personResults.PersonAttribute

                next()
            }
        function (session, results, next) {
            //set the flag         session.conversationData.isPhoneNum . to true if asked for phone number.
            mapPersonAttribute(session, session.conversationData.PersonAttribute);
            next()
        },
        function (session, results, next) {
           //getPersonInfo() gives info about USER from MS graph. 
            personInfoHelper.getPersonInfo(session.userData.authUser.email, session.conversationData.PersonName)
                .then(function (resolvedPerson) {
                    session.conversationData.userInfoRsolvedPerson = resolvedPerson
                    next()
                });
        }, function (session, results, next) {
                session.replaceDialog('singleUserInfo')
            }
        },function (session, results, next) {
            session.endDialog();
        }
    ]).triggerAction({
        matches: 'UserInfo'
    }) 

  //This dialog will output 'Rohit's phone number is 'xx-xxx-xxxxx''
    bot.dialog('singleUserInfo', [
        function (session, results, next) {
            let name = session.conversationData.userInfoRsolvedPerson[0].name
            let phonenumber = session.conversationData.userInfoRsolvedPerson[0].phonenumber
            let title = session.conversationData.userInfoRsolvedPerson[0].title

        
          if (session.conversationData.isPhoneNum) {
                session.say(`${name}'s phone number is ${phonenumber}.`,
                    `${name}'s phone number is ${phonenumber}.`,
                    { inputHint: builder.InputHint.expectingInput }); 
            }else{
                session.say(`Sorry I don't have that specific information about ${name}`,
                    `Sorry I don't have that specific information about ${name}`,
                { inputHint: builder.InputHint.expectingInput });
            }
            next()/*need to have so all flags are cleared.*/
         }, function (session, results, next) {
            
            session.endDialog()
        }
    ]);



}//module ends

function clearFlags(session) {
    session.conversationData.isPhoneNum = false;
    session.conversationData.isTitle = false;
}

function mapPersonAttribute(session, personAttribute) {
   if (personAttribute.includes('phone')) {
        session.conversationData.isPhoneNum = true;
    }else {
        session.replaceDialog('generalErrorNoContext');
    }
}
{code}

can you help to set the test case for this? if it's going to be different that what you mentioned.

**Comment by microsoftly** [comment]
I would love to help, but it would help me to first fix the formatting. Can you make sure you're using the right formatting for your code? It should be wrapped in triple ticks with a javascript note
```
```javascript
// code goes here
``` <- end ticks

I would also recommend just playing around with making simple bots first. Make 1 dialog, add in a prompt, compose multiple dialogs, use user data, etc ....

**Comment by Stevenic** [comment]
@microsoftly Your BotTester is pretty cool. Would love to get your feedback on the way we support testing in the BotBuilder v4 SDK we're working on here:

https://github.com/Microsoft/botbuilder-js

We have a very similar `TestAdapter` class here:

https://github.com/Microsoft/botbuilder-js/blob/master/libraries/botbuilder/src/testAdapter.ts

And you can see examples of us using it in unit tests here:

https://github.com/Microsoft/botbuilder-js/blob/master/libraries/botbuilder/tests/core.js#L41

We're still a few months away from a public preview of v4 so lots of time to influence our architecture.

**Comment by rjgmail88** [comment]
@microsoftly, sorry for the formatting. I just cleaned my last comment.

**Comment by fanidamj** [comment]
Closing the issue assuming it as resolved. Please reopen a new issue if it still persists.
